web-api/src/storage/postgres.rs

```diff
@@ -649,11 +672,81 @@ impl Storage {
         id: &DocumentId,
     ) -> Result<Option<NormalizedEmbedding>, Error> {
         let mut tx = self.postgres.pool.begin().await?;
-        let embedding = Database::get_embedding(&mut tx, id).await?;
+        let mut embedding = Database::get_embedding(&mut tx, id).await?;
         tx.commit().await?;
 
+        if !*self.is_migrated.read().await
+            && embedding
+                .as_deref()
+                .map(|embedding| embedding.is_empty())
+                .unwrap_or_default()
+        {
+            embedding = self.elastic.get_embedding(id).await?;
+        }
+
         Ok(embedding)
     }
+
+    pub(crate) async fn migrate(&self, n: u16) -> Result<bool, Error> {
+        if n == 0 {
+            return Ok(false);
+        }
+
+        let mut tx = self.postgres.pool.begin().await?;
+
+        let ids = sqlx::query_as::<_, DocumentId>(
+            "SELECT document_id
+            FROM document
+            WHERE cardinality(embedding) = 0
+            LIMIT $1
+            FOR UPDATE;",
```

Comment:
This can block a large number of documents for the users for possibly a bit of time since ElasticSearch is involved.
It could be better do one document at a time and parallelize the work instead.

Also, what happens in this case if a write to a property was requested and we are in the case that ElasticSearch has not to be updated yet? I think the property update will overwrite the data with the new value. Is this correct?

Reply:
> This can block a large number of documents for the users for possibly a bit of time since ElasticSearch is involved.

`FOR UPDATE` will only lock `LIMIT` number of documents, the others won't be locked. this is why there are the parameters in the query to migrate `n` documents per time window `t`, eg we could migrate 10docs/1sec which would leave the number of locked documents quite small. this also doesn't mean that the docs are locked for the whole time window, but only for how long it takes to migrate them, basically one pg call to lock the ids, then one es call to get the doc data, then one pg call to insert the data.

> It could be better do one document at a time and parallelize the work instead.

parallelization could also be an issue if we have to many parallel calls to the dbs instead of just a few batched calls.

> Also, what happens in this case if a write to a property was requested and we are in the case that ElasticSearch has not to be updated yet? I think the property update will overwrite the data with the new value. Is this correct?

it is important to note that with the new deployment all insert/delete operations of documents/properties will be done to pg & es, not only es. the fallback only applies to data from before the new deployment. then we can have multiple scenarios:
- a write to a property of a doc which has been ingested after the new deployment or a doc which has already be migrated: the write will happen to pg & es since the migration won't touch it again
- a write to a property of a doc which has not yet been migrated and migration is running: the write can't happen because of the lock
- a write to a property of a doc which has not yet been migrated and migration is not yet running: here we can run into overwrite issues (with old data instead of new data), eg if the following order happens 1) write new property to pg, 2) lock & run doc migration which copies old property from es to pg, 3) write new property to es. i'll add locks to prevent that, we can remove them after the migration.

the last point is essentially an issue caused by the missing transaction system wrapping both es & pg. for example this can also happen for documents with no migration involved. image the following: 1) ingest document which writes to pg, 2) immediately call delete which deletes from pg, 3) delete tries at es but doesn't find the doc yet which is ok for es, 4) ingest doc to es happens to be slow this time and occurs after the delete. similar issues can be constructed for ingesting the doc twice with the same id but different data, which attempts to overwrite in our current implementation, then we could also end up with unmatching pg & es data. there are likely more of those cases, all of them hard to guard against without a transaction system.

Reply:
pushed the lock addtions

Reply:
Ok, I'm convinced, but let's keep these numbers low.

Reply:
I don't thing the locks prevent that.

Stretching the transaction across the ES calls just makes sure that you can't have partial success/failure.

But it doesn't make sure that others reading data from elastic search do get the update before the postgres transaction concludes.

Furthermore to get what we currently get shouldn't it be enough to `FOR UPDATE` lock in the migration query before fetching the properties from elastic search instead of acquiring the lock in (nearly?) every place?

Also I'm not sure anyone used the property updates, wouldn't it have been simpler to deploy a version which stores all new documents in PG already and fetches the old way and has a switch to switch between old-property-updates, no-property-updates, new-property-updates, then disable updates temporary run a separate "scrip" to migrate all documents fastly and then switch over to new updates? Sure that would have been a bit of downtime for (property updates only) but AFIK with how things are currently used that would be much much simpler and reliable then a truly fluent gradual background update. 

@Robert-Steiner do we have statistics about how often the property update endpoints are used, if at all?

Reply:
i'd guess that the properties endpoint is not much used (if at all).

wdyt about the following: under the feature flag new data gets stored already to pg (in addition to es), fetching still behaves the old way from es. then we can run the migration/data copy, without the need to decide where to fetch from for each request. in the following pr which removes the feature flag we switch the fetching from pg to es. i'll push an update later and ping you.

Reply:
seems good :+1:

Reply:
> @Robert-Steiner do we have statistics about how often the property update endpoints are used, if at all?
unfortunately, we currently have no insight into corresponding endpoints

Reply:
i pushed the simplifications, most of the removed code moved into the following pr #833, except for the fallback logic which was removed completely.